File: DFSBFS/미로탐색.py
from collections import deque

N, M = map(int, input().split())
graph = []

for _ in range(N):
    graph.append(list(map(int, input())))


def bfs(x, y, visited = []):
    queue = deque([(x, y)])

    dx = [-1, 1, 0, 0] # 상 하 좌 우
    dy = [0, 0, -1, 1]

    while queue:
        x, y = queue.popleft()
        visited.append((x, y))

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < M and graph[nx][ny] == 1:
                if (nx, ny) not in visited:
                    graph[nx][ny] = graph[x][y] + 1
                    queue.append((nx, ny))

    return graph[N - 1][M - 1]

print(bfs(0, 0))

File: Algorithm/Algorithm/그래프탐색/미로탐색.swift
//
//  main.swift
//  Algorithm
//
//  Created by 이건준 on 2023/10/03.
//

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let N = input[0]
let M = input[1]

var graph: [[Int]] = []

for _ in 0..<N {
  graph.append(readLine()!.map { Int(String($0))! })
}

let dx = [0, 0, -1, 1] // 상 하 좌 우
let dy = [-1, 1, 0, 0]

func bfs(x: Int, y: Int) {
  
  var queue: [(Int, Int)] = [(x, y)]
  
  while !queue.isEmpty {
    let (x, y) = queue.removeFirst()
    
    for i in 0..<4 {
      let rx = dx[i] + x
      let ry = dy[i] + y
      if 0 <= rx && rx < N && 0 <= ry && ry < M && graph[rx][ry] == 1 {
        graph[rx][ry] = graph[x][y] + 1
        queue.append((rx, ry))
      }
    }
  }
}

bfs(x: 0, y: 0)
print(graph[N-1][M-1])
